pkg/trtl/peers.go

```diff
@@ -96,41 +117,57 @@ func (p *PeerService) RmPeers(ctx context.Context, in *peers.Peer) (out *peers.P
 func (p *PeerService) peerStatus(ctx context.Context, in *peers.PeersFilter) (out *peers.PeersList, err error) {
 	// Create the response
 	out = &peers.PeersList{
-		Peers:  make([]*peers.Peer, 0),
-		Status: &peers.PeersStatus{},
+		Peers: make([]*peers.Peer, 0),
+		Status: &peers.PeersStatus{
+			Regions: make(map[string]int64),
+		},
 	}
 
 	// Iterate over all the peers (necessary for both list and status-only)
 	// TODO: filter self from the list?
-	ps := p.store.ListPeers()
+	ps, err := p.db.Iter(nil, options.WithNamespace(NamespacePeers))
+	if err != nil {
+		return nil, err
+	}
 	defer ps.Release()
 
-	// TODO: Implement peer iteration
-	/*
-		for ps.Next() {
-			peer := ps.Peer()
-			if peer == nil {
-				continue
-			}
-
-			out.Status.NetworkSize++
-			out.Status.Regions[peer.Region]++
-
-			// If it's not a status only, get the details for each Peer
-			if !in.StatusOnly {
-				// If we've been asked to filter by region
-				if len(in.Region) > 0 {
-					for _, region := range in.Region {
-						if peer.Region == region {
-							out.Peers = append(out.Peers, peer)
-						}
+	for ps.Next() {
+		// Skip tombstones
+		// TODO: why is Honu returning tombstones in iter?
+		var obj *object.Object
+		if obj, err = ps.Object(); err != nil {
+			log.Error().Err(err).Str("key", string(ps.Key())).Msg("could not retrieve object from db")
+			continue
+		}
+
+		if obj.Tombstone() {
+			continue
+		}
+
+		peer := new(peers.Peer)
+		if err = proto.Unmarshal(obj.Data, peer); err != nil {
+			log.Warn().Err(err).Str("key", string(ps.Key())).Msg("could not unmarshal peer")
+			continue
+		}
+
+		out.Status.NetworkSize++
+		out.Status.Regions[peer.Region]++
+
+		// If it's not a status only, get the details for each Peer
+		if !in.StatusOnly {
+			// If we've been asked to filter by region
+			if len(in.Region) > 0 {
+				for _, region := range in.Region {
+					if peer.Region == region {
+						out.Peers = append(out.Peers, peer)
```

Comment:
Do we need to break out of the loop here? If the request contains duplicate regions then we could end up returning the same peer more than once.

Reply:
Good point, I'll make that change.